### PR TITLE
Feature bit 182 optimize address validation

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs [hunk 1]
```diff
@@ -20,6 +20,23 @@ public class BlackListService : IBlackListService
...
+            await ThrowOnNotSupportedBlockchainType(blockchainType);
+
+            var blackList = await _blackListRepository.TryGetAsync(blockchainType, blockedAddress);
+
+            if (blackList == null)
+            {
+                return false;
+            }
+
+            var isBlocked = blackList.IsCaseSensitive && blockedAddress == blackList.BlockedAddress ||
```
Comment (cassade) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs]: Parentheses?

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainApiClientProvider.cs [hunk 2]
```diff
@@ -1,17 +1,23 @@
...
     {
-        private readonly IIndex<string, IBlockchainApiClient> _clients;
+        private readonly IImmutableDictionary<string, IBlockchainApiClient> _clients;
 
-        public BlockchainApiClientProvider(IIndex<string, IBlockchainApiClient> clients)
+        public BlockchainApiClientProvider(BlockchainsIntegrationSettings settings, ILogFactory logFactory, int blockchainApiTimeoutSeconds)
         {
-            _clients = clients;
+            var timeout = TimeSpan.FromSeconds(blockchainApiTimeoutSeconds);
+            _clients = settings?.Blockchains.ToImmutableDictionary(x => x.Type,
```
Comment (cassade) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainApiClientProvider.cs]: Does _clients field may be null? Later it's used without checks.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            _blockchainConnectAttemptsDelays = new ConcurrentDictionary<string, int>();
+        }
+
+        public void FireInitializationAndForget()
+        {
+            lock (_lock)
+            {
+                if (!_canBeInitialized)
+                {
+                    throw new InvalidOperationException("Can be initialized only once.");
```
Comment (cassade) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs:41]: Does it really may be called multiple times? Where is exception handling logic in such case?

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            {
+                throw new InvalidOperationException($"There is no cache for {blockchainType} yet");
+            }
+
+            return isPublicExtensionRequired;
+        }
+
+        private async Task InitializeOneAsync(string blockchainType, IBlockchainApiClient apiClient)
+        {
+            while (true)
```
Comment (cassade) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs:70]: Maybe it's better to use Polly for retry logic.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs (hunk 1 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs]: image
Redundant parentheses...

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainApiClientProvider.cs (hunk 2 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainApiClientProvider.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs (hunk 3 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs:41]: Caller just can't call it twice and that is nice. 
Also, in current code it is called only once. 
But such check does not make code worse.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs (hunk 4 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlockchainWalletsCacheService.cs:70]: Well, it is a straightforward and clear way to retry till win. No need to change that

**Review by cassade — APPROVED** [review]

**Pull request merged**